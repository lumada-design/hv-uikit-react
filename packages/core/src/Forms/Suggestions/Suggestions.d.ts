import * as React from "react";
import { StandardProps, PopperProps } from "@mui/material";
import { ListValueProp } from "../..";

export type HvSuggestionsClassKey = "root" | "popper" | "list";

export interface HvSuggestionsProps
  extends StandardProps<React.HTMLAttributes<HTMLDivElement>, HvSuggestionsClassKey> {
  /**
   * Whether suggestions is visible.
   */
  expanded?: boolean;
  /**
   * The HTML element Suggestions attaches to.
   */
  anchorEl?: PopperProps["anchorEl"];
  /**
   * Array of { id, label, ...others } values to display in the suggestion list
   */
  suggestionValues?: ListValueProp[];
  /**
   * Function called when suggestion list is closed
   */
  onClose?: (event: React.SyntheticEvent) => void;
  /**
   * Function called when a suggestion is selected
   */
  onSuggestionSelected?: (event: React.MouseEvent, item: ListValueProp) => void;
}

export default function HvSuggestions(props: HvSuggestionsProps): JSX.Element | null;
