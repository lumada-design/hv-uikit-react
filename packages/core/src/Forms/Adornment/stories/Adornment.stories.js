import React, { useState } from "react";

import { Preview, PreviewOff, Success, Fail } from "@hitachivantara/uikit-react-icons";

import { HvBaseInput, HvFormElement, HvLabel, HvAdornment, HvButton } from "../../..";

export default {
  title: "Components/Forms/Adornment",
  parameters: {
    componentSubtitle: null,
    usage: 'import { HvAdornment } from "@hitachivantara/uikit-react-core";',
  },
  component: HvAdornment,
  decorators: [(storyFn) => <div style={{ width: "600px" }}>{storyFn()}</div>],
};

export const Main = () => {
  const inputId = "controlled-input";
  const inputLabelId = "controlled-input-label";
  const [isPassword, setPasswordType] = useState(true);
  const toggleType = () => {
    setPasswordType(!isPassword);
  };
  const getIcon = (isPass) => (isPass ? <Preview /> : <PreviewOff />);

  return (
    <HvFormElement value="p455w0rd" status="standBy">
      <HvLabel id={inputLabelId} label="First name">
        <HvBaseInput
          id={inputId}
          type={isPassword ? "password" : "text"}
          placeholder="Insert your passwords"
          endAdornment={
            <HvAdornment
              aria-label="show password"
              icon={getIcon(isPassword)}
              onClick={toggleType}
            />
          }
        />
      </HvLabel>
    </HvFormElement>
  );
};

export const DynamicAdornments = () => {
  const inputId = "valid-controlled-input";
  const inputLabelId = "valid-controlled-input-label";
  const [status, setStatus] = useState("valid");
  const btnStyle = {
    width: "250px",
    height: "50px",
    margin: "10px",
  };
  return (
    <>
      <HvFormElement value="content" status={status}>
        <HvLabel id={inputLabelId} label="First name">
          <HvBaseInput
            id={inputId}
            placeholder="Insert your name"
            endAdornment={
              <>
                <HvAdornment showWhen="invalid" icon={<Fail semantic="sema4" />} />
                <HvAdornment showWhen="valid" icon={<Success semantic="sema1" />} />
              </>
            }
          />
        </HvLabel>
      </HvFormElement>
      <HvButton
        style={btnStyle}
        onClick={() => {
          setStatus("valid");
        }}
      >
        Set valid
      </HvButton>
      <HvButton
        style={btnStyle}
        onClick={() => {
          setStatus("invalid");
        }}
      >
        Set invalid
      </HvButton>
    </>
  );
};

DynamicAdornments.story = {
  parameters: {
    docs: {
      storyDescription: "Adornment showcasing the ability to react to the form element status.",
    },
  },
};
