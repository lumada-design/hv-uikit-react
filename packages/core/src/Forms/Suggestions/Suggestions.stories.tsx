import { useRef, useState } from "react";
import { Meta, StoryObj } from "@storybook/react";
import {
  HvBaseInput,
  HvBaseInputProps,
  HvFormElement,
  HvLabel,
  HvSuggestions,
  HvSuggestionsProps,
} from "@hitachivantara/uikit-react-core";

import countryList from "../../Input/countries";

const meta: Meta<typeof HvSuggestions> = {
  title: "Guides/Forms/Form Element Blocks/Suggestions",
  component: HvSuggestions,
  decorators: [
    (Story) => <div style={{ width: 500, height: 320 }}>{Story()}</div>,
  ],
};
export default meta;

export const Main: StoryObj<HvSuggestionsProps> = {
  args: {
    expanded: true,
  },
  argTypes: {
    classes: { control: { disable: true } },
  },
  render: () => {
    const suggestions = countryList;
    const [open, setOpen] = useState(false);
    const [suggestionList, setSuggestionList] = useState<string[]>([]);
    const [value, setValue] = useState("");
    const inputRef = useRef<HTMLElement>(null);

    const handleChange: HvBaseInputProps["onChange"] = (e, val) => {
      const matches = suggestions.filter((v) =>
        v.toUpperCase().startsWith(val.toUpperCase()),
      );
      const newList = val.length >= 1 ? matches : [];
      setSuggestionList(newList);
      setOpen(newList.length > 0);
      setValue(val);
    };

    const handleSelection: HvSuggestionsProps["onSuggestionSelected"] = (
      e,
      val,
    ) => {
      console.log(val);
      setOpen(false);
      setValue(val.label as any);
      inputRef?.current?.focus();
    };

    const handleSuggestionsKey: HvSuggestionsProps["onKeyDown"] = (evt) => {
      if (evt.code === "Esc") {
        inputRef?.current?.focus();
        setOpen(false);
      } else if (evt.code === "Tab") {
        if (evt.shiftKey) {
          setTimeout(() => inputRef?.current?.focus());
        } else {
          setOpen(false);
        }
      }
    };

    const values = suggestionList
      .map((label, id) => ({ id: String(id), label }))
      .slice(0, 6);

    return (
      <HvFormElement>
        <HvLabel id="countries" label="Select country">
          <HvBaseInput
            value={value}
            inputRef={inputRef}
            placeholder="Insert country"
            onChange={handleChange}
            inputProps={{ "aria-labelledby": "countries" }}
          />
          <HvSuggestions
            expanded={open}
            anchorEl={inputRef.current?.parentElement}
            onClose={() => setOpen(false)}
            onKeyDown={handleSuggestionsKey}
            onSuggestionSelected={handleSelection}
            suggestionValues={values}
          />
        </HvLabel>
      </HvFormElement>
    );
  },
};

export const ServerSideSuggestions: StoryObj<HvSuggestionsProps> = {
  render: () => {
    const [open, setOpen] = useState(false);
    const [suggestionList, setSuggestionList] = useState<string[]>([]);
    const [value, setValue] = useState("");
    const inputRef = useRef<HTMLElement>(null);

    // Server-side mock function
    const fetchCountries = (input: string) =>
      new Promise((resolve) => {
        const countries = countryList
          .filter((c) => c.toUpperCase().includes(input.toUpperCase()))
          .slice(0, 8);

        setTimeout(() => resolve(countries), 300);
      });

    const handleChange: HvBaseInputProps["onChange"] = (e, val) => {
      setValue(val);
      setOpen(false);
      if (val.length < 1) return;
      fetchCountries(val).then((countries) => {
        setSuggestionList(countries as string[]);
        setOpen((countries as string[]).length >= 1);
      });
    };

    const handleSelection: HvSuggestionsProps["onSuggestionSelected"] = (
      e,
      val,
    ) => {
      console.log(val);
      setOpen(false);
      setValue(val.label as any);
      inputRef?.current?.focus();
    };

    const handleSuggestionsKey: HvSuggestionsProps["onKeyDown"] = (evt) => {
      if (evt.code === "Esc") {
        inputRef?.current?.focus();
        setOpen(false);
      } else if (evt.code === "Tab") {
        if (evt.shiftKey) {
          setTimeout(() => inputRef?.current?.focus());
        } else {
          setOpen(false);
        }
      }
    };

    const values = suggestionList
      .map((label, id) => ({ id: String(id), label }))
      .slice(0, 6);

    return (
      <HvFormElement>
        <HvLabel id="countries" label="Select country">
          <HvBaseInput
            value={value}
            inputRef={inputRef}
            placeholder="Insert country"
            onChange={handleChange}
            inputProps={{ "aria-labelledby": "countries" }}
          />
          <HvSuggestions
            expanded={open}
            anchorEl={inputRef.current?.parentElement}
            onClose={() => setOpen(false)}
            onKeyDown={handleSuggestionsKey}
            onSuggestionSelected={handleSelection}
            suggestionValues={values}
          />
        </HvLabel>
      </HvFormElement>
    );
  },
};

export const OpenWithDownArrow: StoryObj<HvSuggestionsProps> = {
  render: () => {
    const suggestions = countryList;
    const [open, setOpen] = useState(false);
    const [suggestionList, setSuggestionList] = useState<string[]>([]);
    const [value, setValue] = useState("");
    const inputRef = useRef<HTMLElement>(null);

    const handleChange: HvBaseInputProps["onChange"] = (e, val) => {
      const matches = suggestions.filter((v) =>
        v.toUpperCase().startsWith(val.toUpperCase()),
      );
      const newList = val.length >= 1 ? matches : [];
      setSuggestionList(newList);
      setOpen(newList.length > 0);
      setValue(val);
    };

    const handleSelection: HvSuggestionsProps["onSuggestionSelected"] = (
      e,
      val,
    ) => {
      console.log(val);
      setOpen(false);
      setValue(val.label as any);
      inputRef?.current?.focus();
    };

    const handleKey: HvBaseInputProps["onKeyDown"] = (e) => {
      if (e.code === "ArrowDown") {
        setOpen(true);
        document.getElementById("suggestions-list-item-0")?.focus();
      }
    };

    const handleSuggestionsKey: HvSuggestionsProps["onKeyDown"] = (evt) => {
      if (evt.code === "Esc") {
        inputRef?.current?.focus();
        setOpen(false);
      } else if (evt.code === "Tab") {
        if (evt.shiftKey) {
          setTimeout(() => inputRef?.current?.focus());
        } else {
          setOpen(false);
        }
      }
    };

    const values = suggestionList
      .map((label, id) => ({ id: String(id), label }))
      .slice(0, 6);

    return (
      <HvFormElement>
        <HvLabel id="countries" label="Select country">
          <HvBaseInput
            value={value}
            inputRef={inputRef}
            placeholder="Insert country"
            onChange={handleChange}
            onKeyDown={handleKey}
            inputProps={{ "aria-labelledby": "countries" }}
          />
          <HvSuggestions
            id="suggestions"
            expanded={open}
            anchorEl={inputRef.current?.parentElement}
            onClose={() => setOpen(false)}
            onKeyDown={handleSuggestionsKey}
            onSuggestionSelected={handleSelection}
            suggestionValues={values}
          />
        </HvLabel>
      </HvFormElement>
    );
  },
};
