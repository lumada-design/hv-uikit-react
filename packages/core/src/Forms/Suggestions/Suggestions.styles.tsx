import { theme } from "@hitachivantara/uikit-styles";

import { createClasses } from "../../utils/classes";

export const { staticClasses, useClasses } = createClasses("HvSuggestions", {
  root: { position: "relative" },
  list: {
    backgroundColor: theme.colors.atmo1,
    boxShadow: theme.colors.shadow,
    padding: theme.space.xs,
    width: "100%",
  },
  popper: {
    width: "100%",
    position: "absolute",
    transform: "translate3d(0, -1px, 0) !important",
    zIndex: theme.zIndices.tooltip,
  },
});
