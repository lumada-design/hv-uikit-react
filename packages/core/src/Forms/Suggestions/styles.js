const styles = (theme) => ({
  root: {
    position: "relative",
    "& $list": {
      backgroundColor: theme.palette.atmo1,
      boxShadow: theme.hv.shadows[1],
      padding: theme.spacing("sm"),
    },
  },
  popper: {
    width: "100%",
    position: "absolute",
    transform: "translate3d(0, -1px, 0) !important",
    zIndex: theme.zIndex.tooltip,
  },
  list: {},
});

export default styles;
