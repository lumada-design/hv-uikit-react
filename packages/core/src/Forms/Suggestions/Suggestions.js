import React, { useContext } from "react";
import clsx from "clsx";
import PropTypes from "prop-types";
import OutsideClickHandler from "react-outside-click-handler";
import { Popper } from "@mui/material";
import { withStyles } from "@mui/styles";
import { HvSelectionList, HvListItem } from "../..";
import { setId } from "../../utils";
import { HvFormElementContext } from "../FormElement";
import styles from "./styles";

/**
 * Allow text input boxes to trigger the display of one or more predictions of the user's intended value.
 */
const HvSuggestions = (props) => {
  const {
    id,
    className,
    classes,
    expanded = false,
    anchorEl,
    suggestionValues = [],
    onClose,
    onSuggestionSelected,
    ...others
  } = props;

  const { elementId } = useContext(HvFormElementContext);
  const localId = id ?? setId(elementId, "suggestions");

  console.log(classes.popper);

  return (
    <div id={localId} className={clsx(className, classes.root)} {...others}>
      {/* <Popper open={expanded} disablePortal anchorEl={anchorEl} className={classes.popper}>
        <OutsideClickHandler onOutsideClick={(e) => onClose?.(e)}>
          <HvSelectionList
            className={classes.list}
            id={setId(localId, "list")}
            onChange={onSuggestionSelected}
          >
            {suggestionValues?.map((item, i) => {
              const itemKey = item.id || setId("item", i);

              return (
                <HvListItem key={itemKey} value={item} disabled={item.disabled || undefined}>
                  {item.label}
                </HvListItem>
              );
            })}
          </HvSelectionList>
        </OutsideClickHandler>
      </Popper> */}
    </div>
  );
};

HvSuggestions.formElementType = "controlled";

HvSuggestions.propTypes = {
  /**
   * Id to be applied to the root node.
   */
  id: PropTypes.string,
  /**
   * Class names to be applied.
   */
  className: PropTypes.string,
  /**
   * A Jss Object used to override or extend the styles applied.
   */
  classes: PropTypes.shape({
    /**
     * Styles applied to the component root class.
     */
    root: PropTypes.string,
    /**
     * Styles applied to the popper component.
     */
    popper: PropTypes.string,
    /**
     * Styles applied to the suggestion list container.
     */
    list: PropTypes.string,
  }).isRequired,
  /**
   * Whether suggestions is visible.
   */
  expanded: PropTypes.bool,
  /**
   * The HTML element Suggestions attaches to.
   */
  anchorEl: PropTypes.oneOfType([PropTypes.element, PropTypes.object, PropTypes.func]),
  /**
   * Array of { id, label, ...others } values to display in the suggestion list
   */
  suggestionValues: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      label: PropTypes.node.isRequired,
      value: PropTypes.any,
    })
  ),
  /**
   * Function called when suggestion list is closed
   */
  onClose: PropTypes.func,
  /**
   * Function called when a suggestion is selected
   */
  onSuggestionSelected: PropTypes.func,
};

export default withStyles(styles, { name: "HvSuggestions" })(HvSuggestions);
