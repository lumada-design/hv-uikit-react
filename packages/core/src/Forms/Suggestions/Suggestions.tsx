import React, {
  forwardRef,
  useContext,
  useEffect,
  useRef,
  useState,
} from "react";

import MuiPopper from "@mui/material/Popper";
import { useForkRef } from "@mui/material/utils";

import { HvBaseProps } from "@core/types/generic";
import { setId } from "@core/utils/setId";
import { HvListItem } from "@core/ListContainer";
import {
  HvClickOutsideEvent,
  useClickOutside,
} from "@core/hooks/useClickOutside";
import { ExtractNames } from "@core/utils/classes";
import { HvSelectionList } from "@core/SelectionList";

import { staticClasses, useClasses } from "./Suggestions.styles";
import { HvFormElementContext } from "../FormElement";

export { staticClasses as suggestionsClasses };

export type HvSuggestionsClasses = ExtractNames<typeof useClasses>;

export interface HvSuggestion {
  id?: string;
  label: React.ReactNode;
  value?: string;
  disabled?: boolean;
}

export interface HvSuggestionsProps extends HvBaseProps {
  /** Whether suggestions is visible. */
  expanded?: boolean;
  /** The HTML element Suggestions attaches to. */
  anchorEl?: HTMLElement | null;
  /** Array of { id, label, ...others } values to display in the suggestion list */
  suggestionValues?: HvSuggestion[] | null;
  /** Function called when a suggestion is selected */
  onSuggestionSelected?: (event: React.MouseEvent, value: HvSuggestion) => void;
  /** Function called when suggestion list is closed */
  onClose?: (event: HvClickOutsideEvent) => void;
  /** A Jss Object used to override or extend the styles applied to the component. */
  classes?: HvSuggestionsClasses;
}

export const HvSuggestions = forwardRef((props: HvSuggestionsProps, extRef) => {
  const {
    id,
    className,
    classes: classesProp,
    expanded = false,
    anchorEl,
    suggestionValues = [],
    onClose,
    onSuggestionSelected,
    ...others
  } = props;
  const { classes, cx } = useClasses(classesProp);

  const { elementId } = useContext(HvFormElementContext);
  const localId = id ?? setId(elementId, "suggestions");

  const ref = useRef<HTMLDivElement>(null);
  const forkedRef = useForkRef(ref, extRef);

  const [isOpen, setIsOpen] = useState(expanded);

  useClickOutside(ref, (event) => {
    setIsOpen(false);
    onClose?.(event);
  });

  useEffect(() => {
    setIsOpen(expanded);
  }, [expanded]);

  return (
    <div
      id={localId}
      ref={forkedRef}
      className={cx(classes.root, className)}
      {...others}
    >
      <MuiPopper
        open={isOpen}
        disablePortal
        anchorEl={anchorEl}
        className={classes.popper}
      >
        <HvSelectionList
          className={classes.list}
          id={setId(localId, "list")}
          onChange={onSuggestionSelected}
        >
          {suggestionValues?.map((item, i) => {
            const itemKey = item.id || setId("item", i);

            return (
              <HvListItem
                key={itemKey}
                value={item}
                disabled={item.disabled || undefined}
              >
                {item.label}
              </HvListItem>
            );
          })}
        </HvSelectionList>
      </MuiPopper>
    </div>
  );
});
