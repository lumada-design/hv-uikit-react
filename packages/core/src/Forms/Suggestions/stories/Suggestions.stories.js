import React, { useState, useRef } from "react";
import parser from "html-react-parser";
import { useTheme } from "@mui/styles";

import {
  HvFormElement,
  HvLabel,
  HvBaseInput,
  HvButton,
  HvSuggestions,
  HvTypography,
} from "../../..";
import { KeyboardCodes, isKeypress } from "../../../utils";

import countryList from "../../../Input/stories/countries";

export default {
  title: "How To Guides/Forms/Form Element Blocks/Suggestions",
  parameters: {
    componentSubtitle: null,
    usage: 'import { HvSuggestions } from "@hitachivantara/uikit-react-core"',
  },
  component: HvSuggestions,
  decorators: [
    (Story) => (
      <div style={{ width: 500, height: 320 }}>
        <Story />
      </div>
    ),
  ],
};

const { Esc, Tab } = KeyboardCodes;

export const Main = () => {
  const suggestions = countryList;
  const [open, setOpen] = useState(false);
  const [suggestionList, setSuggestionList] = useState([]);
  const [value, setValue] = useState("");
  const inputRef = useRef(null);

  const handleChange = (e, val) => {
    const matches = suggestions.filter((v) => v.toUpperCase().startsWith(val.toUpperCase()));
    const newList = val.length >= 1 ? matches : [];
    setSuggestionList(newList);
    setOpen(newList.length > 0);
    setValue(val);
  };

  const handleSelection = (e, val) => {
    console.log(val);
    setOpen(false);
    setValue(val.label);
    inputRef?.current?.focus();
  };

  const handleSuggestionsKey = (evt) => {
    if (isKeypress(evt, Esc)) {
      inputRef?.current?.focus();
      setOpen(false);
    } else if (isKeypress(evt, Tab)) {
      if (evt.shiftKey) {
        setTimeout(() => inputRef?.current?.focus());
      } else {
        setOpen(false);
      }
    }
  };

  const values = suggestionList.map((label, id) => ({ id: String(id), label })).slice(0, 6);

  return (
    <HvFormElement>
      <HvLabel id="countries" label="Select country">
        <HvBaseInput
          value={value}
          inputRef={inputRef}
          placeholder="Insert country"
          onChange={handleChange}
          inputProps={{ "aria-labelledby": "countries" }}
        />
        <HvSuggestions
          expanded={open}
          anchorEl={inputRef.current?.parentElement}
          onClose={() => setOpen(false)}
          onKeyDown={handleSuggestionsKey}
          onSuggestionSelected={handleSelection}
          suggestionValues={values}
        />
      </HvLabel>
    </HvFormElement>
  );
};

export const ServerSideSuggestions = () => {
  const [open, setOpen] = useState(false);
  const [suggestionList, setSuggestionList] = useState([]);
  const [value, setValue] = useState("");
  const inputRef = useRef(null);

  // Server-side mock function
  const fetchCountries = (input) =>
    new Promise((resolve) => {
      const countries = countryList
        .filter((c) => c.toUpperCase().includes(input.toUpperCase()))
        .slice(0, 8);

      setTimeout(() => resolve(countries), 300);
    });

  const handleChange = (e, val) => {
    setValue(val);
    setOpen(false);
    if (val.length < 1) return;
    fetchCountries(val).then((countries) => {
      setSuggestionList(countries);
      setOpen(countries.length >= 1);
    });
  };

  const handleSelection = (e, val) => {
    console.log(val);
    setOpen(false);
    setValue(val.label);
    inputRef?.current?.focus();
  };

  const handleSuggestionsKey = (evt) => {
    if (isKeypress(evt, Esc)) {
      inputRef?.current?.focus();
      setOpen(false);
    } else if (isKeypress(evt, Tab)) {
      if (evt.shiftKey) {
        setTimeout(() => inputRef?.current?.focus());
      } else {
        setOpen(false);
      }
    }
  };

  const values = suggestionList.map((label, id) => ({ id: String(id), label })).slice(0, 6);

  return (
    <HvFormElement>
      <HvLabel id="countries" label="Select country">
        <HvBaseInput
          value={value}
          inputRef={inputRef}
          placeholder="Insert country"
          onChange={handleChange}
          inputProps={{ "aria-labelledby": "countries" }}
        />
        <HvSuggestions
          expanded={open}
          anchorEl={inputRef.current?.parentElement}
          onClose={() => setOpen(false)}
          onKeyDown={handleSuggestionsKey}
          onSuggestionSelected={handleSelection}
          suggestionValues={values}
        />
      </HvLabel>
    </HvFormElement>
  );
};

export const OpenWithDownArrow = () => {
  const suggestions = countryList;
  const [open, setOpen] = useState(false);
  const [suggestionList, setSuggestionList] = useState([]);
  const [value, setValue] = useState("");
  const inputRef = useRef(null);

  const handleChange = (e, val) => {
    const matches = suggestions.filter((v) => v.toUpperCase().startsWith(val.toUpperCase()));
    const newList = val.length >= 1 ? matches : [];
    setSuggestionList(newList);
    setOpen(newList.length > 0);
    setValue(val);
  };

  const handleSelection = (e, val) => {
    console.log(val);
    setOpen(false);
    setValue(val.label);
    inputRef?.current?.focus();
  };

  const handleKey = (e) => {
    if (isKeypress(e, KeyboardCodes.ArrowDown)) {
      setOpen(true);
      document.getElementById("suggestions-list-item-0")?.focus();
    }
  };

  const handleSuggestionsKey = (evt) => {
    if (isKeypress(evt, Esc)) {
      inputRef?.current?.focus();
      setOpen(false);
    } else if (isKeypress(evt, Tab)) {
      if (evt.shiftKey) {
        setTimeout(() => inputRef?.current?.focus());
      } else {
        setOpen(false);
      }
    }
  };

  const values = suggestionList.map((label, id) => ({ id: String(id), label })).slice(0, 6);

  return (
    <HvFormElement>
      <HvLabel id="countries" label="Select country">
        <HvBaseInput
          value={value}
          inputRef={inputRef}
          placeholder="Insert country"
          onChange={handleChange}
          onKeyDown={handleKey}
          inputProps={{ "aria-labelledby": "countries" }}
        />
        <HvSuggestions
          id="suggestions"
          expanded={open}
          anchorEl={inputRef.current?.parentElement}
          onClose={() => setOpen(false)}
          onKeyDown={handleSuggestionsKey}
          onSuggestionSelected={handleSelection}
          suggestionValues={values}
        />
      </HvLabel>
    </HvFormElement>
  );
};

export const WithHighlighter = () => {
  const suggestions = countryList;
  const [open, setOpen] = useState(false);
  const [suggestionList, setSuggestionList] = useState([]);
  const [value, setValue] = useState("");
  const [mode, setMode] = useState("auto-complete");
  const inputRef = useRef(null);
  const theme = useTheme();

  const filterHighlighter = (searchString) => {
    const reg = new RegExp(searchString, "gi");
    switch (mode) {
      case "auto-complete":
        return suggestions
          .filter((v) => v.toUpperCase().startsWith(searchString.toUpperCase()))
          .map((match) => ({
            value: match,
            label: parser(`${match.replace(reg, "<b>$&</b>")}`),
          }));

      case "auto-suggest":
        return suggestions
          .filter((v) => v.toUpperCase().includes(searchString.toUpperCase()))
          .map((match) => ({
            value: match,
            label: parser(`<b>${match.replace(reg, "</b>$&<b>")}</b>`),
          }));

      default:
        return searchString;
    }
  };

  const handleChange = (e, val) => {
    const newList = val.length >= 1 ? filterHighlighter(val) : [];
    setSuggestionList(newList);
    setOpen(newList.length > 0);
    setValue(val);
  };

  const handleSelection = (e, val) => {
    console.log(val);
    setOpen(false);
    setValue(val.value);
    inputRef?.current?.focus();
  };

  const handleSuggestionsKey = (evt) => {
    if (isKeypress(evt, Esc)) {
      inputRef?.current?.focus();
      setOpen(false);
    } else if (isKeypress(evt, Tab)) {
      if (evt.shiftKey) {
        setTimeout(() => inputRef?.current?.focus());
      } else {
        setOpen(false);
      }
    }
  };

  const values = suggestionList
    .map((elem, id) => ({ id: String(id), label: elem.label, value: elem.value }))
    .slice(0, 4);

  return (
    <>
      <HvTypography>
        {`Current mode: `}
        <b>{mode}</b>
      </HvTypography>
      <HvButton
        style={{ margin: theme.hvSpacing("xs", 0, "md") }}
        onClick={() => setMode(mode === "auto-complete" ? "auto-suggest" : "auto-complete")}
      >
        Change mode
      </HvButton>
      <p />
      <HvFormElement>
        <HvLabel id="countries" label="Select country">
          <HvBaseInput
            value={value}
            inputRef={inputRef}
            placeholder="Insert country"
            onChange={handleChange}
            inputProps={{ "aria-labelledby": "countries" }}
          />
          <HvSuggestions
            expanded={open}
            anchorEl={inputRef.current?.parentElement}
            onClose={() => setOpen(false)}
            onKeyDown={handleSuggestionsKey}
            onSuggestionSelected={handleSelection}
            suggestionValues={values}
          />
        </HvLabel>
      </HvFormElement>
    </>
  );
};
