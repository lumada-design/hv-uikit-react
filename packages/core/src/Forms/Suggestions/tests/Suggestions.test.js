/* eslint-env jest */

import React from "react";
import { mount } from "enzyme";
import { Main } from "../stories/Suggestions.stories";
import { HvProvider, HvSuggestions } from "../../..";

describe("v3 Suggestions", () => {
  let wrapper;

  it("should be defined", () => {
    wrapper = mount(
      <HvProvider>
        <Main />
      </HvProvider>
    );
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(wrapper.find(HvSuggestions)).toMatchSnapshot();
  });

  it("should render the Suggestions", () => {
    const component = wrapper.find(HvSuggestions);
    expect(component.length).toBe(1);
  });
});
