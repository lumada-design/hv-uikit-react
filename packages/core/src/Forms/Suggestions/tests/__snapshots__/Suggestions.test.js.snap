// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suggestions should render correctly 1`] = `
<WithStyles(HvSuggestions)
  expanded={false}
  onClose={[Function]}
  onKeyDown={[Function]}
  onSuggestionSelected={[Function]}
  suggestionValues={Array []}
>
  <HvSuggestions
    classes={
      Object {
        "list": "HvSuggestions-list",
        "popper": "HvSuggestions-popper",
        "root": "HvSuggestions-root",
      }
    }
    expanded={false}
    onClose={[Function]}
    onKeyDown={[Function]}
    onSuggestionSelected={[Function]}
    suggestionValues={Array []}
  >
    <div
      className="HvSuggestions-root"
      id="hvformelement1-suggestions"
      onKeyDown={[Function]}
    />
  </HvSuggestions>
</WithStyles(HvSuggestions)>
`;
